lib/property_constraints.js

```diff
@@ -1,95 +1,116 @@
 var Immutable = require('immutable');
 
-var cardinality = {
-  '=': function(val, count) {
-    return val.size === parseInt(count);
-  },
-  '>': function(val, count) {
-    return val.size > parseInt(count);
-  },
-  '<': function(val, count) {
-    return val.size < parseInt(count);
-  }
+function not(fn) {
+  return function(state) {
+    return !fn(state);
+  };
+}
+
+function and() {
+  var fns = Array.prototype.slice.call(arguments);
+  return function(state) {
+    return fns.reduce(function(memo, fn) {
+      return memo && fn(state);
+    }, true);
+  };
+}
+
+
+var constraint = function() {
```

Comment:
The thing is, this is not exclusively a constraint. Any predicate over state is a constraint.

I'd like to use the terminology from https://leanpub.com/javascriptallongesix/read as much as possible, this if folks need to know how the combinators stuff works we can just link them there. Without building our own universe of terminology.

In general I'm thinking any of the dsl will be an opt-in additional lib. Folks should understand things are just predicates over state before they get fancy ways to build them.